pkg/crd/spec.go

```diff
@@ -36,6 +39,102 @@ type SpecMarker interface {
 	ApplyToCRD(crd *apiext.CustomResourceDefinitionSpec, version string) error
 }
 
+type VersionType string
+
+const (
+	GA    VersionType = "ga"
+	Beta  VersionType = "beta"
+	Alpha VersionType = "alpha"
+)
+
+type VersionName string
+
+// Value creates a float64 type of value to represent a VersionName
+func (v VersionName) Value() float64 {
+	// remove the first character, `v`
+	name := string(v)[1:]
+	name = strings.Replace(name, "beta", ".", -1)
+	name = strings.Replace(name, "alpha", ".", -1)
+	result, _ := strconv.ParseFloat(name, 64)
+	return result
+}
+
+// VersionNames implements sort.Interface for []VersionName based on VersionName.Value
+// all the items in VersionNames should have the same VersionType
+type VersionNames []VersionName
+
+func (v VersionNames) Len() int           { return len(v) }
+func (v VersionNames) Swap(i, j int)      { v[i], v[j] = v[j], v[i] }
+func (v VersionNames) Less(i, j int) bool { return v[i].Value() < v[j].Value() }
+
+// Strings creates a slice of string where each string corresponds to one VersionName
+func (v VersionNames) Strings() []string {
+	result := []string{}
+	for _, item := range v {
+		result = append(result, string(item))
+	}
+	return result
+}
+
+func sortVersionNames(versionNames []string) []string {
+	patterns := map[VersionType]string{
```

Comment:
this can be generalized into `^v([0-9]+)((alpha|beta)([0-9]*))?$`, at which point you can numerically sort groups 1 and 4, and string (lexigraphically) sort group 3, IIRC.  The regex looks a bit more complicated, but the sorting logic is less complicated

Reply:
removed